zui/src/ui/defaultextension.ts

```diff
@@ -0,0 +1,184 @@
+import { z } from 'zod'
+import { UICategorySchema, UIComponentDefinitions } from './types'
+import { ComponentImplementationMap } from './types'
+
+const commonInputSchema = z.object({
+  label: z.string().optional(),
+  readonly: z.boolean().optional(),
+})
+
+export const defaultExtensions = {
+  string: {
+    textbox: {
+      id: 'textbox',
+      schema: commonInputSchema.extend({
+        multiline: z.boolean().default(false).optional(),
+        fitContentWidth: z.boolean().default(false).optional(),
+      }),
+    },
+    datetimeinput: {
+      id: 'datetimeinput',
+      schema: commonInputSchema.extend({
+        type: z.enum(['time', 'date', 'date-time']).default('date-time'),
+      }),
+    },
+  },
+  number: {
+    numberinput: {
+      id: 'numberinput',
+      schema: commonInputSchema,
+    },
+    slider: {
+      id: 'slider',
+      schema: commonInputSchema,
+    },
+  },
+  boolean: {
+    checkbox: {
+      id: 'checkbox',
+      schema: commonInputSchema,
+    },
+  },
+  array: {
+    select: {
+      id: 'select',
+      schema: z.undefined(),
+    },
+  },
+  object: {
+    verticalLayout: {
+      id: 'verticalLayout',
+      schema: z.object({}),
+    },
+    horizontalLayout: {
+      id: 'horizontalLayout',
+      schema: z.object({}),
+    },
+    group: {
+      id: 'group',
+      schema: z.object({
+        label: z.string(),
+      }),
+    },
+    categorization: {
+      id: 'categorization',
+      schema: z.object({}),
+    },
+    category: {
+      id: 'category',
+      schema: z.object({
+        label: z.string(),
+      }),
+    },
+  },
+} as const satisfies UIComponentDefinitions
+
+export const defaultExtensionComponents: ComponentImplementationMap<typeof defaultExtensions> = {
+  string: {
+    textbox: (_, __, { multiline, label, fitContentWidth, readonly }, { scope }) => {
+      return {
+        type: 'Control',
+        scope,
+        label: label,
+        options: {
+          multi: multiline,
+          readOnly: readonly,
+          trim: fitContentWidth,
+        },
+      }
+    },
+    datetimeinput: (_, __, { type, label, readonly }, { scope }) => {
+      return {
+        type: 'Control',
+        scope,
+        label,
+        options: {
+          format: type,
+          readOnly: readonly,
+        },
+      }
+    },
+  },
+  number: {
+    numberinput: (_, __, { readonly }, { scope }) => {
+      return {
+        type: 'Control',
+        scope,
+        options: {
+          readOnly: readonly,
+        },
+      }
+    },
+    slider: (_, __, { label, readonly }, { scope }) => {
```

Comment:
currently when implementing a component the function is `(type, id, params, context) => UISchema`, I think I should move `type` and `id` to `context` now that it is less used

Reply:
Moved!